src/lock.rs

```diff
@@ -0,0 +1,174 @@
+            self.name,
+            other.name
+        );
+        ensure!(
+            self.checksum == other.checksum,
+            "Checksum mismatch - expected {}, actual {}",
+            self.checksum,
+            other.checksum
+        );
+        ensure!(
+            self.repository == other.repository,
+            "Repository mismatch - expected {}, actual {}",
+            self.repository,
+            other.repository
+        );
+        ensure!(
+            self.version == other.version,
+            "Version mismatch - expected {}, actual {}",
+            self.version,
+            other.version
+        );
+        ensure!(
+            self.dependencies == other.dependencies,
+            "Dependencies mismatch"
+        );
+        Ok(())
+    }
+
+    /// Constructs a Dependency instance with matching metadata
+    pub fn as_dependency(&self) -> Dependency {
+        Dependency {
+            package: self.name.clone(),
+            manifest: DependencyManifest {
+                version: VersionReq {
+                    comparators: vec![semver::Comparator {
+                        op: semver::Op::Exact,
+                        major: self.version.major,
+                        minor: Some(self.version.minor),
+                        patch: Some(self.version.patch),
+                        pre: self.version.pre.clone(),
+                    }],
+                },
+                repository: self.repository.clone(),
+            },
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+struct RawLockfile {
+    packages: Vec<LockedPackage>,
+}
+
+/// Captures metadata about currently installed Packages
+///
+/// Used to ensure future installations will deterministically select the exact same packages.
+#[derive(Default)]
+pub struct Lockfile {
+    packages: HashMap<PackageName, Arc<LockedPackage>>,
+}
+
+impl Lockfile {
+    /// Checks if the Lockfile currently exists in the filesystem
+    pub async fn exists() -> eyre::Result<bool> {
+        fs::try_exists(LOCKFILE)
+            .await
+            .wrap_err("Failed to detect lockfile")
+    }
+
+    /// Loads the Lockfile from the current directory
+    pub async fn read() -> eyre::Result<Self> {
+        let toml = fs::read_to_string(LOCKFILE)
+            .await
+            .wrap_err("Failed to read lockfile")?;
+
+        let raw: RawLockfile = toml::from_str(&toml).wrap_err("Failed to parse lockfile")?;
+
+        Ok(Self::from_iter(raw.packages.into_iter()))
+    }
+
+    /// Loads the Lockfile from the current directory, if it exists, otherwise returns an empty one
+    pub async fn read_or_default() -> eyre::Result<Self> {
+        if Lockfile::exists().await? {
+            Lockfile::read().await
+        } else {
+            Ok(Lockfile::default())
+        }
+    }
+
+    /// Persists a Lockfile to the filesystem
+    pub async fn write(&self) -> eyre::Result<()> {
+        let raw = RawLockfile {
+            packages: self
```

Comment:
it'd be nice here for any lockfile to be _canonical_ - that there's only one valid written down version of a lockfile. This ensures that random changes in implementation don't lead to massive diffs.

Reply:
Oh yeah, good point. I was going to sort stuff to make it deterministic but forgot about it. Will fix tomorrow.

Reply:
note - would be good to sort internally (e.g. on list of dependencies) as well as global list.